ext/drbbdup/drbbdup.c

```diff
@@ -0,0 +1,1198 @@
+    }
+
+    /* Also take into account default case. */
+    if (manager->default_case.is_defined)
+        count++;
+
+    return count;
+}
+
+/**
+ *  Clone from original instrlist, but place duplication in bb.
+ */
+static void
+drbbdup_add_copy(void *drcontext, instrlist_t *bb, instrlist_t *orig_bb)
+{
+    if (instrlist_first(orig_bb) != NULL) {
+        instrlist_t *dup = instrlist_clone(drcontext, orig_bb);
+        instr_t *start = instrlist_first(dup);
+        instrlist_prepend(bb, start);
+
+        /* Empty list and destroy. Do not use clear as instrs are needed. */
+        instrlist_init(dup);
+        instrlist_destroy(drcontext, dup);
+    }
+}
+
+/* Creates a manager, which contains book-keeping data for a fragment. */
+static drbbdup_manager_t *
+drbbdup_create_manager(void *drcontext, void *tag, instrlist_t *bb)
+{
+    drbbdup_manager_t *manager = dr_global_alloc(sizeof(drbbdup_manager_t));
+    memset(manager, 0, sizeof(drbbdup_manager_t));
+
+    manager->cases = NULL;
+    ASSERT(opts.dup_limit > 0, "dup limit should be greater than zero");
+    manager->cases = dr_global_alloc(sizeof(drbbdup_case_t) * opts.dup_limit);
+    memset(manager->cases, 0, sizeof(drbbdup_case_t) * opts.dup_limit);
+    manager->enable_dup = true;
+    manager->ref_counter = 1;
+
+    ASSERT(opts.set_up_bb_dups != NULL, "set up call-back cannot be NULL");
+    manager->default_case.encoding = opts.set_up_bb_dups(
+        manager, drcontext, tag, bb, &(manager->enable_dup), opts.user_data);
+
+    /* Check whether user wants copies for this particular bb. */
+    if (!manager->enable_dup && manager->cases != NULL) {
+        /* Multiple cases not wanted. Destroy cases. */
+        dr_global_free(manager->cases, sizeof(drbbdup_case_t) * opts.dup_limit);
+        manager->cases = NULL;
+    }
+
+    manager->default_case.is_defined = true;
+    return manager;
+}
+
+/* Transforms the bb to contain additional copies (within the same fragment. */
+static void
+drbbdup_set_up_copies(void *drcontext, instrlist_t *bb, drbbdup_manager_t *manager)
+{
+    ASSERT(manager != NULL, "manager should not be NULL");
+    ASSERT(manager->enable_dup, "bb duplication should be enabled");
+    ASSERT(manager->cases != NULL, "cases should not be NULL");
+
+    /**
+     *  Example: Lets say we have the following bb:
+     *   mov ebx ecx
+     *   mov esi eax
+     *   ret
+     *
+     * We require 2 cases, we need to construct the bb as follows:
+     *   LABEL 1
+     *   mov ebx ecx
+     *   mov esi eax
+     *   jmp EXIT LABEL
+     *
+     *   LABEL 2
+     *   mov ebx ecx
+     *   mov esi eax
+     *   EXIT LABEL
+     *   ret
+     *
+     * The inclusion of the dispatcher is left for the instrumentation stage.
+     *
+     * Note, we add jmp instructions here and DR will set them to meta automatically.
+     */
+
+    /* We create a duplication here to keep track of original bb. */
+    instrlist_t *original = instrlist_clone(drcontext, bb);
+
+    /**
```

Comment:
style: Do not use Doxygen comment style for non-doc comments.  Ditto at several other places in this file.

Reply:
Done.